bst.py

```diff
@@ -83,13 +107,76 @@ def balance(self):
         negative balance if the right side is deeper, and zero if
         the tree is balanced
         """
-        if self.left is None and self.right is None:
-            return 0
-        if self.left is None:
-            return -1 * self.right.depth()
-        if self.right is None:
-            return self.left.depth()
-        return self.left.depth() - self.right.depth()
+        l = self.left.depth() if self.left else 0
+        r = self.right.depth() if self.right else 0
+        return l - r
+
+    # Depth first traverals
+    def pre_order(self):
+        """
+        Call a pre-order depth first traversal generator.
+        Return nodes in turn from generator.
+        """
+        parent_stack = []
+        while parent_stack or self is not None:
+            if self is not None:
+                yield self
+                # self is the symbol for the currently traversed node
+                if self.right is not None:
+                    parent_stack.append(self.right)
+                self = self.left
+            else:
+                self = parent_stack.pop()
+
+    def in_order(self):
+        """
+        Call an in-order depth first traversal generator.
+        Return nodes in turn from generator.
+        """
+        parent_stack = []
+        while parent_stack or self is not None:
+            if self is not None:
+                parent_stack.append(self)
+                self = self.left
+            else:
+                self = parent_stack.pop()
+                yield self
+                self = self.right
+
+    def post_order(self):
+        """
+        Call a post-order depth first traversal generator.
+        Return nodes in turn from generator.
+        """
+        parent_stack = []
+        last_node = None
+        while parent_stack or self is not None:
+            if self is not None:
+                parent_stack.append(self)
+                self = self.left
+            else:
```

Comment:
This will never get called.. This is saying if self is None.. But if it is None.. It will never enter the while loop..